### PR TITLE
feat: Support multiple verifications keys

### PR DESCRIPTION
**Pull request by Michaelhobo**
https://hub.mender.io/t/multiple-artifactverifykeys/4309/6

Creates a new client config parameter called ArtifactVerifyKeys that is a list of paths to keys. Any matching key can be used to verify an artifact - e.g. if 5 verification keys are provided, only 1 needs to match to verify the artifact.

This is version 2, see https://github.com/mendersoftware/mender/pull/1045 for previous discussions.

Signed-off-by: Michael Ho <callmemikeh@gmail.com>

### CODE REVIEW

**Comment by Michaelhobo** [comment]
@kacf @estenberg I finally had time to fix up https://github.com/mendersoftware/mender/pull/1045

I made the requested changes - we now accept a list ArtifactVerifyKeys that is a repeated version of ArtifactVerifyKey, and it will return an error if both are set.

File: vendor/github.com/mendersoftware/mender-artifact/areader/reader.go [hunk 1]
```diff
@@ -397,10 +385,19 @@ func (ar *Reader) readHeaderV3(version []byte) error {
...
+				shouldBeSigned = ar.ShouldBeSignedCallback(ar.hInfo)
+			}
 			// Artifact should be signed, but isn't, so do not process the update.
-			if ar.shouldBeSigned && !ar.IsSigned {
+			if shouldBeSigned && !ar.IsSigned {
 				return errors.New("reader: expecting signed artifact, but no signature file found")
 			}
+			// First read and verify signature
+			if err = signatureReadAndVerify(ar.signature, ar.manifest.GetRaw(),
+				ar.VerifySignatureCallback, shouldBeSigned); err != nil {
```
Comment (kacf) [inline vendor/github.com/mendersoftware/mender-artifact/areader/reader.go:398]: If the signature verification is done here, then if I read it correctly, it will be done multiple times, once for each file in the tar (likely only header and payload at the moment, but could increase if we implement multiple payloads). Do you actually need to move it out of its original location?

The same goes for the V2 variant.

File: vendor/github.com/mendersoftware/mender-artifact/areader/reader.go [hunk 2]
```diff
@@ -34,6 +34,7 @@ import (
 )
 
 type SignatureVerifyFn func(message, sig []byte) error
+type ShouldBeSignedFn func(artifact.HeaderInfoer) bool
```
Comment (kacf) [inline vendor/github.com/mendersoftware/mender-artifact/areader/reader.go:37]: For simplicity and review purposes we can keep these changes here now, but before this is merged, all vendor files needs to be submitted as a separate pull request in https://github.com/mendersoftware/mender-artifact, and then they need to be update using the go mod vendor mechanism.

File: installer/dual_rootfs_device_test.go [hunk 3]
```diff
@@ -20,6 +20,7 @@ import (
 	"testing"
 	"time"
 
+	"github.com/mendersoftware/mender/conf"
```
Comment (kacf) [inline installer/dual_rootfs_device_test.go:23]: This file needs its copyright year updated at the very top.

File: app/standalone.go [hunk 4]
```diff
@@ -113,7 +113,7 @@ func doStandaloneInstallStatesDownload(art io.ReadCloser, key []byte,
 		// No doStandaloneFailureStates here, since we have not done anything yet.
 		return nil, err
 	}
-	installer, installers, err := installer.ReadHeaders(art, dt, key,
+	installer, installers, err := installer.ReadHeaders(art, dt, device.Config.GetVerificationKeys(),
```
Comment (kacf) [inline app/standalone.go:116]: Linter is complaining that this line is too long; just put the last argument on a new line.

**Review by kacf — CHANGES_REQUESTED** [review]
Thanks @Michaelhobo!

* Remember that you need `Changelog` and `Ticket` tags as described [here](https://github.com/mendersoftware/mender/blob/master/CONTRIBUTING.md#structuring-your-commits). In the changelog tag, instead of one of the predefined strings it might be useful to describe which configuration option is being added and how it works.

* Please add a pull request in [mender-docs](https://github.com/mendersoftware/mender-docs/blob/master/03.Client-installation/07.Configuration-file/50.Configuration-options/docs.md#artifactverifykey) as well, to document this new option.

File: vendor/github.com/mendersoftware/mender-artifact/areader/reader.go (hunk 2 above)
Comment (Michaelhobo) [reply vendor/github.com/mendersoftware/mender-artifact/areader/reader.go:37]: Sounds good. I'm not familiar with the standard go tools (I usually use bazel in a monorepo), could I get some pointers? I see `# github.com/mendersoftware/mender-artifact v0.0.0-20220913084855-9ed8ad0d53d0` in modules.txt - how do I generate a new version for this when we're ready?

File: vendor/github.com/mendersoftware/mender-artifact/areader/reader.go (hunk 1 above)
Comment (Michaelhobo) [reply vendor/github.com/mendersoftware/mender-artifact/areader/reader.go]: You're right, I originally did this because we didn't now if the payload needed to be signed (and with what key) until after reading the header. Since we removed the key scoping, we can roll this back to the original.

File: vendor/github.com/mendersoftware/mender-artifact/areader/reader.go (hunk 2 above)
Comment (Michaelhobo) [reply vendor/github.com/mendersoftware/mender-artifact/areader/reader.go]: I ended up removing all my areader changes, so I no longer need this. But, I do have another PR that is in a similar situation (https://github.com/Michaelhobo/mender-artifact-1/pull/2 ), so I'd like some pointers anyway.

File: vendor/github.com/mendersoftware/mender-artifact/areader/reader.go [hunk 5]
```diff
@@ -575,6 +573,16 @@ func (ar *Reader) readHeaderV2(version []byte) error {
 		if err := ar.readHeader(hc, comp); err != nil {
 			return err
 		}
+		shouldBeSigned := ar.ShouldBeSignedCallback(ar.hInfo)
+		// Artifact should be signed, but isn't, so do not process the update.
+		if shouldBeSigned && !ar.IsSigned {
+			return errors.New("reader: expecting signed artifact, but no signature file found")
+		}
```
[the remark this answers is not in the document]
Comment (Michaelhobo) [reply vendor/github.com/mendersoftware/mender-artifact/areader/reader.go]: Reverted all this

**Review by Michaelhobo — COMMENTED** [review]
I updated the changelog/ticket, I think my new commit description is reasonable for a changelog, so I just put `All`.

I will make the mender-docs PR shortly

**Comment by Michaelhobo** [comment]
mender-docs PR is here: https://github.com/mendersoftware/mender-docs/pull/1975

File: installer/dual_rootfs_device_test.go (hunk 3 above)
Comment (kacf) [reply installer/dual_rootfs_device_test.go:23]: Hehe, shrewd. The year is compared against the author date of the latest commit, so you might need `git commit --amend --reset-author` to reset it in 2023.

**Review by kacf — APPROVED** [review]
Thanks @Michaelhobo, now it looks compact and ready to go!

The copyright year now fails because of 2023, but I'll run it through our full CI in the meantime, and we can fix that next week.

Happy New Year! :fireworks:

**Comment by kacf** [comment]
The failure in CI is due to a known unrelated issue which we can ignore. So I think this should be good to go.

Please run `git commit --amend --reset-author` in order to update the author date to 2023, the rest of the CI should pass then.

**Comment by Michaelhobo** [comment]
Thanks for the review @kacf , should be ready to go.

**Comment by kacf** [comment]
Already tested in https://gitlab.com/Northern.tech/Mender/mender-qa/-/pipelines/734757667, will not rerun since nothing has changed.

**Comment by kacf** [comment]
Thanks @Michaelhobo!
